pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,320 @@
+)
+
+type ruleType int
+
+type ruleEvent uint8
+
+const (
+	unicast   ruleType = 0
+	igmp      ruleType = 1
+	multicast ruleType = 2
+
+	ruleAdd    ruleEvent = 1
+	ruleUpdate ruleEvent = 2
+	ruleDelete ruleEvent = 3
+
+	groupAddressIndex = "groupAddressIndex"
+)
+
+type mcastItem struct {
+	groupAddress net.IPNet
+	ruleIDs      map[string]*uint16
+}
+
+type mcastIGMPRule struct {
+	ruleID       string
+	priority     *uint16
+	groupAddress []string
+	event        ruleEvent
+}
+
+// multicastController handls only IGMP protocol, it stores the group address to RuleID map and
+// also rule to group addresses
+type multicastController struct {
+	ofClient openflow.Client
+
+	ifaceStore interfacestore.InterfaceStore
+	ruleCache  *ruleCache
+	queue      workqueue.RateLimitingInterface
+	// key is multicast group address
+	mcastItemRuleCache cache.Indexer
+
+	//key is ruleID while the value is list of group addresses
+	ruleIDGroupAddressMap map[string]sets.String
+	ruleIDGroupMapMutex   sync.RWMutex
+}
+
+func groupAddressKey(obj interface{}) (string, error) {
+	group := obj.(*mcastItem)
+	return group.groupAddress.IP.String(), nil
+}
+
+func groupAddressRuleIndexFunc(obj interface{}) ([]string, error) {
+	item := obj.(*mcastItem)
+	ruleIDs := make([]string, len(item.ruleIDs))
+	for ruleID := range item.ruleIDs {
+		ruleIDs = append(ruleIDs, ruleID)
+	}
+	return ruleIDs, nil
+}
+
+func (c *multicastController) addGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleAdd,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) updateGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleUpdate,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) deleteGroupAddressForTableIDs(ruleID string, groupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		groupAddress: groupAddresses,
+		event:        ruleDelete,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) validation(iface *interfacestore.InterfaceConfig,
+	groupAddress net.IPNet, direction crdv1beta.Direction) (*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error) {
+
+	var ruleTypePtr *crdv1beta.NetworkPolicyType
+	action, uuid, ruleName := v1alpha1.RuleActionAllow, apitypes.UID(""), ""
+	if iface == nil {
+		// check the if iface exists
+		klog.InfoS("Iface should not be empty")
+		return nil, "", nil, "", fmt.Errorf("iface should not be empty")
+	}
+	ns, podname := iface.PodNamespace, iface.PodName
+	obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress.String())
+	if !exists {
+		obj, exists, _ = c.mcastItemRuleCache.GetByKey(groupAddress.IP.String())
```

Comment:
I don't think it is useful to use the IP as key to query, since you always transfer the IP to a CIDR with "/32" mask when adding object into the cache?

Reply:
done, change to ip address without mask.